ostd/src/task/atomic_mode.rs

```diff
@@ -45,3 +48,39 @@ pub fn might_sleep() {
         );
     }
 }
+
+/// A marker trait for guard types that enforce the atomic mode.
+///
+/// Key kernel primitives such as `SpinLock` and `Rcu` rely on
+/// [the atomic mode](crate::task::atomic_mode).
+/// The existence of such a guard guarantees that
+/// the current task is executing in the atomic mode.
+///
+/// Such guard types include [`DisabledLocalIrqGuard`] and [`DisabledPreemptGuard`].
+/// Users can also implement this trait for additional guard types via
+/// [`BorrowAsAtomicModeGuard`] safely.
+///
+/// # Safety
+///
+/// The implementer must ensure that the atomic mode is maintained while
+/// the guard type is alive.
+pub unsafe trait AtomicModeGuard {}
+
+// SAFETY: The guard disables local IRQs, which meets the first
+// sufficient condition for atomic mode.
+unsafe impl AtomicModeGuard for DisabledLocalIrqGuard {}
+// SAFETY: The guard disables preemptions, which meets the second
+// sufficient condition for atomic mode.
+unsafe impl AtomicModeGuard for DisabledPreemptGuard {}
```

Comment:
Let's avoid inter-dependencies between modules. The concept of atomic mode is more fundamental. Let the `guard` module to rely on the `atomic_mode` module (not the other way around) and implement the traits.